drivers/azure/azureutil/azureutil.go

```diff
@@ -104,33 +104,73 @@ func (a AzureClient) resourceGroupExists(ctx context.Context, name string) (bool
 	return checkResourceExistsFromError(err)
 }
 
-// CreateNetworkSecurityGroup either creates or updates the definition of the requested security group with the specified rules and adds it to our DeploymentContext
-func (a AzureClient) CreateNetworkSecurityGroup(ctx context.Context, deploymentCtx *DeploymentContext, resourceGroup, name, location string, rules *[]network.SecurityRule) error {
-	log.Info("Configuring network security group.", logutil.Fields{
-		"name":     name,
-		"location": location})
-	securityGroupsClient := a.securityGroupsClient()
-	future, err := securityGroupsClient.CreateOrUpdate(ctx, resourceGroup, name,
-		network.SecurityGroup{
+// ConfigureNetworkSecurityGroup creates or gets a NSG, updates the configuration, and adds it to our deploymentCtx
+func (a AzureClient) ConfigureNetworkSecurityGroup(ctx context.Context, deploymentCtx *DeploymentContext, resourceGroup, resourceID, location string,
+	usedInPool bool, sshPort, dockerPort int, extraPorts []string) error {
+	// Parse the resource ID to extract values
+	resource, err := azure.ParseResourceID(resourceID)
+	if err != nil {
+		return fmt.Errorf("unable to parse resource ID of network security group: %s", resourceID)
+	}
+	f := logutil.Fields{
+		"nsg":        resource,
+		"usedInPool": usedInPool,
+	}
+	log.Info("Configuring network security group.", f)
+	nsg, err := a.securityGroupsClient().Get(ctx, resource.ResourceGroup, resource.ResourceName, "")
+	exists, err := checkResourceExistsFromError(err)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		// NOTE(aiyengar2): resources should only be created within the driver's provided resource group; however,
+		// an outside NSG can be referenced
+		if resource.ResourceGroup != resourceGroup {
+			return fmt.Errorf("cannot create a security group outside of provided resource group %s: %s",
+				resourceGroup, f)
+		}
+		log.Info("Creating a new security group", f)
+		nsg = network.SecurityGroup{
 			Location: to.StringPtr(location),
 			SecurityGroupPropertiesFormat: &network.SecurityGroupPropertiesFormat{
-				SecurityRules: rules,
+				SecurityRules: &[]network.SecurityRule{},
 			},
-		})
+			// NOTE(aiyengar2): this tag is added to verify that a pooled NSG is safe to be deleted by Rancher
+			// once it is orphaned. If a pooled NSG was not created by Rancher, it will never be deleted.
+			Tags: RancherMachineTags{
+				ManagedByRancher: true,
+			}.CreateTags(),
+		}
+	}
+	if to.String(nsg.Location) != location {
+		return fmt.Errorf("cannot attach security group in %s to a machine provisioned in %s: %s", to.String(nsg.Location), location, f)
+	}
```

Comment:
Without this check, the machine just fails to provision upon trying to create a network interface with the following error below:

![incorrect_location_azure_error](https://user-images.githubusercontent.com/13127443/79923014-96de6f80-83e9-11ea-80e5-9954b0b013f4.png)

We could remove this check, but if so the NSG's `CreateOrUpdate` call will successfully update the NSG in the wrong location with Rancher rules before failing at the `CreateNetworkInterface` call. 

Since there is no code path for removing Rancher rules from a user provided NSG once it has been added (i.e. an NSG that is not ManagedByRancher requires manual cleanup or a reassessment), I think that this should be avoided.

Current error:
![incorrect_location_rancher_error](https://user-images.githubusercontent.com/13127443/79924043-bd050f00-83eb-11ea-85f5-76bab252c500.png)

Reply:
Since we no longer mutate the NSG provided as of https://github.com/rancher/machine/pull/77#issuecomment-625564555, this check should no longer be required; we should fall back to Azure's logic for determining that the NSG is in the wrong location.